### PR TITLE
add usgs gap analysis

### PR DESCRIPTION
**Pull request by amsnyder**
Code to produce gap analysis plots for USGS NWIS data. This should be able to run on NOAA data produced by the munge step with just a few additions to the config.

### CODE REVIEW

**Comment by amsnyder** [comment]
I'm just going to merge this so I can make some more updates--it won't impact any of the scripts you are currently working with.  Let me know if you want to make changes though, and I can make any needed updates!

**Review by galengorski — CHANGES_REQUESTED** [review]
This is looking good, I think these will be really helpful. I got it running on my machine after manually altering one of the munged csv files (see #39). I added some line comments to this PR, other than that, I had a few overall comments:

- Can you make a subdirectory within `data_exploration/out` where the gap analysis plots live? That way they are separate from the gap analysis csvs and a little cleaner to look through.

- I'm looking at the discharge gap plots and the way it is plotted now, it looks like there is missing data in the latest year. Looking at the gap_analysis.csv file, it looks like there is no data for 2021? Is this right or are we just not querying data for 2021?

- I'm looking at the file `usgs_nwis_discharge_01463500_gap_analysis.csv` and I notice a couple of things. For years 2018 and 2020 for example, the p_coverage is 0.939726 but the n_gaps for both years is 0, also for 2012 p_coverage = 1 and n_gap = 1 as well. Seems like maybe the gaps aren't being calculated or logged correctly?

File: utils.py [hunk 1]
```diff
@@ -16,3 +16,13 @@ def prep_write_location(write_location, aws_profile):
     session = boto3.Session(profile_name=aws_profile)
     s3_client = session.client('s3')
     return s3_client
+
```
[the remark this answers is not in the document]
Comment (amsnyder) [reply utils.py:19]: Thanks for catching that!

File: data_exploration/src/gap_analysis.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    var_codes = config['params_to_process']
+    # change to machine+human-readable names
+    var_names = [utils.usgs_nwis_param_code_to_name(code) for code in var_codes]
+
+    # import site:color mapping from config
+    with open("data_exploration/data_exploration_config.yaml", 'r') as stream:
+        config = yaml.safe_load(stream)['gap_analysis.py']
+    # read in data source we want to do gap analysis for
+    source = config['source']
+os.makdirs('data_exploration/out/', exist_ok  = True)
```
Comment (amsnyder) [inline data_exploration/src/gap_analysis.py:88]:
```suggestion
    os.makdirs('data_exploration/out/', exist_ok  = True)
```

File: data_exploration/src/gap_analysis.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+                # year_df.dropna(inplace=True)
+                var_site_gap_df.loc[year, 'p_coverage'] = year_df.count()/365                
+                deltas = year_df.dropna().index.to_series().diff()[1:]
+                gaps = deltas[deltas > dt.timedelta(days=1)]
+                var_site_gap_df.loc[year, 'n_gaps'] = len(gaps)
+                var_site_gap_df.loc[year, 'gap_median_days'] = gaps.median().days if pd.notna(gaps.median().days) else 0
+                var_site_gap_df.loc[year, 'gap_max_days'] = gaps.max().days if pd.notna(gaps.max().days) else 0
+                var_site_gap_df.to_csv(os.path.join('data_exploration', 'out', f'{source}_{var}_{site}_gap_analysis.csv'))
+            metric_dfs[var][site]= var_site_gap_df
+    return metric_dfs
```
Comment (amsnyder) [inline data_exploration/src/gap_analysis.py:54]:
```suggestion
    return metric_dfs, metrics
```

File: data_exploration/src/gap_analysis.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    with open("data_exploration/data_exploration_config.yaml", 'r') as stream:
+        config = yaml.safe_load(stream)['gap_analysis.py']
+    # read in data source we want to do gap analysis for
+    source = config['source']
+    os.makdirs('data_exploration/out/', exist_ok  = True)
+    # fetch site data and compile into nested dictionary of dataframes
+    var_dfs = compile_data(var_names, source)
+
+    # calculate gap metrics per site variable combo
+    metric_dfs = gap_analysis_calc(source, var_dfs)
```
Comment (amsnyder) [inline data_exploration/src/gap_analysis.py:93]:
```suggestion
    metric_dfs, metrics = gap_analysis_calc(source, var_dfs)
```

File: data_exploration/src/gap_analysis.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+    # plot gap metrics per variable
+    site_colors = config[f'{source}_site_colors']
+    plot_gap_analysis(source, metric_dfs, site_colors)
```
Comment (amsnyder) [inline data_exploration/src/gap_analysis.py:97]:
```suggestion
    plot_gap_analysis(source, metric_dfs, metrics, site_colors)
```

File: data_exploration/src/gap_analysis.py [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+                deltas = year_df.dropna().index.to_series().diff()[1:]
+                gaps = deltas[deltas > dt.timedelta(days=1)]
+                var_site_gap_df.loc[year, 'n_gaps'] = len(gaps)
+                var_site_gap_df.loc[year, 'gap_median_days'] = gaps.median().days if pd.notna(gaps.median().days) else 0
+                var_site_gap_df.loc[year, 'gap_max_days'] = gaps.max().days if pd.notna(gaps.max().days) else 0
+                var_site_gap_df.to_csv(os.path.join('data_exploration', 'out', f'{source}_{var}_{site}_gap_analysis.csv'))
+            metric_dfs[var][site]= var_site_gap_df
+    return metric_dfs
+
+def plot_gap_analysis(source, metric_dfs, site_colors):
```
Comment (amsnyder) [inline data_exploration/src/gap_analysis.py:56]:
```suggestion
def plot_gap_analysis(source, metric_dfs, metrics, site_colors):
```

**Comment by amsnyder** [comment]
For the first bullet point, I added one directory for plots and one for csvs. Let me know if that looks ok.

**Comment by amsnyder** [comment]
For the third bullet point, I was losing gaps at the start and end of the year, and I've updated the code so those are not missed.

**Comment by amsnyder** [comment]
@galengorski - I think I've addressed your concerns, but please let me know if I've missed anything! Otherwise, I think I will merge this request.

**Comment by galengorski** [comment]
Thanks @amsnyder, this sound good to me, I'm not going to run it on my machine, but since it isn't connected to any other part of the pipeline we can merge it now and always revisit if we want to make changes.

**Pull request merged**